### PR TITLE
move `zig cc`, `zig translate-c`, `zig libc`, main(), and linking from stage1 to stage2

### PR DESCRIPTION
**Pull request by andrewrk**
* build.zig: repair the ability to link against llvm, clang, and lld
 * move the zig cc arg parsing logic to stage2
   - the preprocessor flag is still TODO
   - the clang arg iterator code is improved to use slices instead of
     raw pointers because it no longer has to deal with an extern
     struct.
 * clean up error printing with a `fatal` function and use log API
   for messages rather than std.debug.print
 * add support for more CLI options to stage2 & update usage text
   - hooking up most of these new options is TODO
 * clean up the way libc and libc++ are detected via command line
   options. target information is used to determine if any of the libc
   candidate names are chosen.
 * add native library directory detection
 * implement the ability to invoke clang from stage2
 * introduce a build_options.have_llvm so we can comptime branch
   on whether LLVM is linked in or not.

----

Part of the motivation for doing this is so that @alexnask has the option to utilize our mingw .def files and zig cc capabilities to produce .lib files such as kernel32.lib and ntdll.lib for the purposes of testing PE files. This PR is progress towards #4313 and #4314.

### Checklist

 * [x] debug the invalid LLVM IR generated when trying to build this branch
 * [x] ~expose stage2 in stage1 with `zig stage2` for example `zig stage2 build-exe hello.zig`~ better yet, move main() to stage2
 * [x] in Module, utilize std.cache_hash for the root source file
 * [x] add support for invoking clang to build `c_source_files`, integrated with caching system
 * [x] self-host link.cpp and building libcs (#4313 and #4314). using the `zig cc` command will set a flag indicating a preference for the llvm backend, which will include linking with LLD. At least for now. If zig's self-hosted linker ever gets on par with the likes of ld and lld, we can make it always be used even for zig cc.
 * [x] make sure zig cc works and fix preprocessing
 * [x] self-host main.cpp
 * [x] go through the branch diff and look for TODO and add more checklist items. We need this to be on par with master branch before merging so that it doesn't regress.
 * [x] look over the branch diff and open issues for things so we don't lose track of some of the things left to improve.
 * [x] update print_targets.zig with glibc support
 * [x] use global zig-cache dir for crt files
 * [ ] ~avoid invoking lld when it's just 1 object file (the `zig cc -c` case)~ will file separate issue
 * [x] use separate cache hash instances for the zig module and each C object
 * [x] capture lld stdout/stderr better
 * [x] glibc .so files
 * [x] musl
 * [x] mingw-w64
 * [x] retain cache_hash locks on files until we no longer will depend on their presence
 * [x] port the stage1 os.cpp code that raises the open fd limit
 * [ ] ~improve the stage2 tests to support testing with LLVM extensions enabled~ will file separate issue
 * [x] `zig translate-c`
 * [x] `zig libc`
 * [x] ELF LLD linking
 * [x] MachO LLD linking
 * [x] COFF LLD linking
 * [x] WASM LLD linking
 * [x] skip LLD caching when bin directory is not in the cache (so we don't put `id.txt` into the cwd)
 * [x] ability to produce archive file

### CODE REVIEW

File: src-self-hosted/link.zig [hunk 1]
```diff
@@ -16,14 +16,36 @@ pub const Options = struct {
...
+    strip: bool = false,
+    /// If this is true then this link code is responsible for outputting an object
+    /// file and then using LLD to link it together with the link options and other objects.
+    /// Otherwise (depending on `use_llvm`) this link code directly outputs and updates the final binary.
+    use_lld: bool = false,
+    /// If this is true then this link code is responsible for making an LLVM IR Module,
+    /// outputting it to an object file, and then linking that together with link options and
+    /// other objects.
+    /// Otherwise (depending on `use_lld`) this link code directly outputs and updates the final binary.
+    use_llvm: bool = false,
```
Comment (daurnimator) [inline src-self-hosted/link.zig:37]: thoughts on making this an enum: `code_generation: enum { builtin, llvm }` (I'm thinking about how to plumb in GCC support)

**Comment by andrewrk** [comment]
Some stats on the zig executable (release build, stripped):
 - master branch stage1: 119 MiB
 - master branch stage2 (no LLVM): 3.5 MiB
 - this branch stage1 (includes stage2): 128 MiB
 - this branch stage2 (yes LLVM): 107 MiB
 - this branch stage2 (no LLVM): 3.7 MiB

So that 107 number is what we're looking at once we are done self-hosting, with an llvm-enabled build.

Thee 128 number is what we'll be shipping for zig 0.7.0. I do expect it to go down a little bit after deleting the C++ code that was ported from stage1 to stage2 (currently it is duplicated).

The 3.7 number is what you get if you choose to leave LLVM out of your life.

**Comment by bfredl** [comment]
@andrewrk how much does stage2 without llvm do? Debug builds or some optimizations as well?

**Comment by andrewrk** [comment]
You can have a look at some of the test cases to get an idea of what it's capable of so far:

https://github.com/ziglang/zig/tree/master/test/stage2

We have a head start on many features however. Consider the following:
 * `zig fmt` is already self-hosted
 * translate-c / `@cImport` is already self-hosted
 * parsing .d files is already self-hosted
 * the cache hash system is already self-hosted
 * the CLI progress bar thing is already self-hosted
 * detecting libc installation paths is already self-hosted
 * clang CLI arg iteration is already self-hosted

The amount of stage1 C++ code is getting smaller and smaller. This PR takes a huge chunk out of it and moves it to stage2.

**Comment by bfredl** [comment]
I was just trying to build it and it builds fine. But I have just a minor "optical"  issue when building this with cmake (using gmake as backend). The build appears to be "stuck" with `Scanning dependencies of target zig_build_zig1` as the last message from cmake. It builds just fine if I give it a few mins, but I would guess a message like "building target zig_build_zig1" would be more appropriate, as I'm sure it is not just scanning dependencies for several minutes.

**Comment by andrewrk** [comment]
Unfortunately I think that's a cmake feature request rather than something fixable on our end

File: src-self-hosted/Compilation.zig [hunk 2]
```diff
@@ -1484,26 +1688,19 @@ pub fn classifyFileExt(filename: []const u8) FileExt {
...
-    // Look for .so.X, .so.X.Y, .so.X.Y.Z
-    var it = mem.split(filename, ".");
-    _ = it.next().?;
-    var so_txt = it.next() orelse return .unknown;
-    while (!mem.eql(u8, so_txt, "so")) {
-        so_txt = it.next() orelse return .unknown;
+    } else if (mem.endsWith(u8, filename, ".zig")) {
+        return .zig;
+    } else if (mem.endsWith(u8, filename, ".zir")) {
+        return .zig;
```
Comment (Rocknest) [inline src-self-hosted/Compilation.zig:1694]: Did you mean zir?

Comment (andrewrk) [reply src-self-hosted/Compilation.zig:1694]: Sure did, thanks!

**Comment by bfredl** [comment]
@andrewrk cmake should already allow a message for custom object, but you need to specify it explicitly. This seems to work:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1a19c275d..11dbd2980 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -463,6 +463,7 @@ if("${ZIG_TARGET_TRIPLE}" STREQUAL "native")
       COMMAND zig0 ${BUILD_ZIG1_ARGS}
       DEPENDS zig0
       BYPRODUCTS "${ZIG1_OBJECT}"
+      COMMENT STATUS "Building zig object ${ZIG1_OBJECT}"
       WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}"
   )
   set(ZIG_EXECUTABLE "${zig_BINARY_DIR}/zig")
```

**Comment by andrewrk** [comment]
> @andrewrk cmake should already allow a message for custom object, but you need to specify it explicitly. This seems to work:

Brilliant, thanks!

File: src/main.zig [hunk 3]
```diff
@@ -1516,7 +1516,7 @@ pub fn buildOutputType(
     var repl_buf: [1024]u8 = undefined;
 
     while (watch) {
-        try stderr.print("🦎 ", .{});
+        try stderr.print("(zig) ", .{});
```
Comment (Rocknest) [inline src/main.zig:1540]: Sad to see him go. Maybe detect unicode support?

**Comment by andrewrk** [comment]
Before merging... 15.1% C++.

image

File: src/mingw.zig [hunk 4]
```diff
@@ -277,6 +279,233 @@ fn add_cc_args(
...
+
+    const final_def_basename = try std.fmt.allocPrint(arena, "{s}.def", .{lib_name});
+    const def_final_path = try comp.global_cache_directory.join(arena, &[_][]const u8{
+        "o", &digest, final_def_basename,
+    });
+
+    const target_def_arg = switch (target.cpu.arch) {
+        .i386 => "-DDEF_I386",
+        .x86_64 => "-DDEF_X64",
+        .arm, .armeb => switch (target.cpu.arch.ptrBitWidth()) {
```
Comment (LemonBoy) [inline src/mingw.zig:345]:
```zig
.arm, .armeb, .thumb, .thumbeb, .aarch64_32 => "-DDEF_ARM32",
.aarch64, .aarch64_be => "-DDEF_ARM64",
```

Not sure about `aarch64_32`, but this follows the original code in `link.cpp`

**Comment by mattnite** [comment]
https://img1.mashed.com/img/gallery/heres-what-happens-when-you-drink-orange-juice-every-day/intro-1587655828.jpg

File: src/link/MachO.zig [hunk 5]
```diff
@@ -282,6 +288,368 @@ pub fn flushModule(self: *MachO, comp: *Compilation) !void {
...
+    }
+    if (self.base.options.bind_global_refs_locally) {
+        try argv.append("-Bsymbolic");
+    }
+
+    if (self.base.options.verbose_link) {
+        Compilation.dump_argv(argv.items);
+    }
+
+    // TODO allocSentinel crashed stage1 so this is working around it.
```
Comment (LemonBoy) [inline src/link/MachO.zig:570]: #6370 is now fixed, `allocSentinel` should work just fine now

Comment (andrewrk) [reply src/link/MachO.zig:570]: Nice!

**Comment by andrewrk** [comment]
@LemonBoy (or anyone else who cares) I'm calling it a night. Feel free to make some commits against this branch if you want, you won't be clobbering any of my work for about 10 hours. At this point it's just fixing the regressions and getting the CI passing. The big list of todo items I just deleted from that BRANCH_TODO file are things I will be filing separate issues for upon merging.

**Comment by FireFox317** [comment]
While trying to debug, the annoying issue #3825 is coming up. Basically `zig test` is exiting without any message and it is really annoying to debug. By doing `gdb --args zig test ...` and then adding a `catch syscall exit_group` I was able to get a backtrace and show that LLD is failing, but not printing anything helpful :(

**Comment by FireFox317** [comment]
There seems to be a general problem when building crt files.
When building crt object files, I think (don't know for sure) that there is no need for LLD to do something, however still lld tries to link the crt files.

On Linux this is doing the following:
`lld -r -error-limit=0 -m elf_i386 -static -o /home/timon/.cache/zig/o/c6b308e548ff9f4708bd3c2cef1d3937/crti.o /home/timon/.cache/zig/o/2141c41c43c3d8761b2ae2d7360f97f1/crti.o --allow-shlib-undefined`. You can see that the input file is an object file and the output file is the same object file. Is this doing anything useful?

The same thing happends on mingw crt and causing an LLD error. `lld -r -ERRORLIMIT:0 -NOLOGO -DEBUG -MACHINE:X86 -OUT:/home/timon/.cache/zig/o/454e11964f32b765f3134841c713b3c1/crt2.o /home/timon/.cache/zig/o/d2d7d738aa8475e8d7add82e3ef0e29d/crt2.o -NODEFAULTLIB -ENTRY:WinMainCRTStartup`

Backtrace:
```gdb
#0  0x00007ffff7aa40f1 in _exit () from /usr/lib/libc.so.6
#1  0x00005555591e4778 in lld::exitLld(int) ()
#2  0x00005555591e9c63 in lld::ErrorHandler::fatal(llvm::Twine const&) ()
#3  0x0000555558e8d7d4 in lld::fatal(llvm::Twine const&) ()
#4  0x0000555559092f08 in lld::coff::LinkerDriver::link(llvm::ArrayRef<char const*>) ()
#5  0x0000555559093937 in lld::coff::link(llvm::ArrayRef<char const*>, bool, llvm::raw_ostream&, llvm::raw_ostream&) ()
#6  0x0000555556d6be07 in ZigLLDLink (oformat=ZigLLVM_COFF, args=0x55555de56a28, arg_count=10, 
    append_diagnostic=0x55555646f0a0 <link.Coff.append_diagnostic>, context_stdout=0x7fffffff57a0, 
    context_stderr=0x7fffffff57d0) at /home/timon/dev/zig/zig-source/src/zig_llvm.cpp:1061
#7  0x0000555556444a7c in link.Coff.linkWithLLD (self=0x55555de6f260, comp=0x55555de81b38)
    at /home/timon/dev/zig/zig-source/src/link/Coff.zig:1139
#8  0x000055555643b1c4 in link.Coff.flush (self=0x55555de6f260, comp=0x55555de81b38)
    at /home/timon/dev/zig/zig-source/src/link/Coff.zig:733
#9  0x00005555563f7b9a in link.File.flush (base=0x55555de6f260, comp=0x55555de81b38)
    at /home/timon/dev/zig/zig-source/src/link.zig:340
#10 0x00005555563f0e65 in Compilation.update (self=0x55555de81b38)
    at /home/timon/dev/zig/zig-source/src/Compilation.zig:1043
#11 0x00005555565a0ac2 in Compilation.updateSubCompilation (sub_compilation=0x55555de81b38)
    at /home/timon/dev/zig/zig-source/src/Compilation.zig:2388
#12 0x000055555659d060 in Compilation.build_crt_file (comp=0x55555de694f8, root_name=..., 
    output_mode=Obj, c_source_files=...) at /home/timon/dev/zig/zig-source/src/Compilation.zig:2843
--Type <RET> for more, q to quit, c to continue without paging--
#13 0x000055555642b3ce in mingw.buildCRTFile (comp=0x55555de694f8, crt_file=crt2_o)
    at /home/timon/dev/zig/zig-source/src/mingw.zig:42
#14 0x00005555563f63ed in Compilation.performAllTheWork (self=0x55555de694f8)
    at /home/timon/dev/zig/zig-source/src/Compilation.zig:1252
#15 0x00005555563f0c67 in Compilation.update (self=0x55555de694f8)
    at /home/timon/dev/zig/zig-source/src/Compilation.zig:1021
#16 0x000055555639a327 in main.updateModule (gpa=0x55555db76c98 <c_allocator_state>, 
    comp=0x55555de694f8, zir_out_path=..., hook=...)
    at /home/timon/dev/zig/zig-source/src/main.zig:1752
#17 0x000055555633783e in main.buildOutputType (gpa=0x55555db76c98 <c_allocator_state>, 
    arena=0x7fffffffdb88, all_args=..., arg_mode=...)
    at /home/timon/dev/zig/zig-source/src/main.zig:1628
#18 0x0000555556311814 in main.mainArgs (gpa=0x55555db76c98 <c_allocator_state>, arena=0x7fffffffdb88, 
    args=...) at /home/timon/dev/zig/zig-source/src/main.zig:132
#19 0x00005555563112eb in main (argc=11, argv=0x7fffffffddc8)
    at /home/timon/dev/zig/zig-source/src/stage1.zig:42
```

File: src/link/Coff.zig [hunk 6]
```diff
@@ -0,0 +1,1203 @@
...
+
+    const is_obj = self.base.options.output_mode == .Obj;
+
+    // Create an LLD command line and invoke it.
+    var argv = std.ArrayList([]const u8).init(self.base.allocator);
+    defer argv.deinit();
+    // Even though we're calling LLD as a library it thinks the first argument is its own exe name.
+    try argv.append("lld");
+    if (is_obj) {
+        try argv.append("-r");
```
Comment (FireFox317) [inline src/link/Coff.zig:884]: On the windows linker (`lld-link`) -r is ignored:
`lld: warning: ignoring unknown argument '-r'`

**Pull request merged**